iree/hal/buffer_caching.c

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/buffer_caching.h"
+
+#include "iree/base/api.h"
+#include "iree/base/tracing.h"
+#include "iree/hal/allocator.h"
+#include "iree/hal/allocator_caching.h"
+
+extern const iree_hal_buffer_vtable_t iree_hal_caching_buffer_vtable;
+
+static iree_hal_caching_buffer_t* iree_hal_caching_buffer_cast(
+    iree_hal_buffer_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_caching_buffer_vtable);
+  return (iree_hal_caching_buffer_t*)base_value;
+}
+
+iree_status_t iree_hal_caching_buffer_wrap(
```

Comment:
@stellaraccident I've uploaded a new patch with my latest changes wherein I added custom buffer type and wrap the underlying buffer with it. I hit a buffer incompatibility issue at (https://github.com/google/iree/blob/main/iree/hal/buffer.c#L695)
```
../iree/hal/buffer.c:181: PERMISSION_DENIED; buffer memory type is not compatible with the requested operation; buffer has HOST_COHERENT, operation requires HOST_VISIBLE; parsing value '1x5xf32=1,-2,-3,4,-5'
```
I tried to triage the issue further but I've not been successful. The program errors out even before it reaches caching buffer's map_range. 

Sorry to come to you with an issue like this, it would be very helpful if you could take a look at this. Thank you!

Reply:
That means your buffer does not have the right memory type bits (as the message says, there's HOST_COHERENT set but not HOST_VISIBLE as required by iree_hal_buffer_map_range - to map memory it needs to be at least HOST_VISIBLE).

Reply:
I tried to wrap the underlying buffer (cuda buffer in this example) using custom type preserving all buffer properties including that of memory type. I was expecting that a call to `iree_hal_buffer_map_range` would invoke `iree_hal_caching_buffer_map_range` which in turn would call the underlying buffer's map_range (https://github.com/google/iree/blob/main/iree/hal/cuda/cuda_buffer.c#L86) which deals with different memory types.

However, when I run an example the program errors out with the memory type mismatch error before hitting either map_range methods (iree_hal_caching_buffer_map_range and iree_hal_cuda_buffer_map_range). I'm not sure if it's due to my wrapping buffer incorrectly.